src/main/java/org/jabref/logic/ai/embeddings/EmbeddingsGenerationTaskManager.java

```diff
@@ -0,0 +1,202 @@
+
+/**
+ * This class manages a queue of embedding generation tasks for one {@link BibEntry} in a {@link org.jabref.model.database.BibDatabase}.
+ * <p>
+ * {@link org.jabref.model.database.BibDatabase} will listen for entries updates and add them to the queue. This class
+ * will start a background task for generating the embeddings for each entry in the queue.
+ * <p>
+ * This class also provides an ability to prioritize the entry in the queue. But it seems not to work well.
+ */
+public class EmbeddingsGenerationTaskManager extends BackgroundTask<Void> {
+    private final BibDatabaseContext databaseContext;
+    private final FilePreferences filePreferences;
+    private final AiService aiService;
+    private final TaskExecutor taskExecutor;
+
+    private final AiIngestor aiIngestor;
+
+    private final List<LinkedFile> linkedFileQueue = new ArrayList<>();
+    private int numOfProcessedFiles = 0;
+
+    private final Object lock = new Object();
+    private boolean isRunning = false;
+    private boolean isBlockingNewTasks = false;
+
+    public EmbeddingsGenerationTaskManager(BibDatabaseContext databaseContext, FilePreferences filePreferences, AiService aiService, TaskExecutor taskExecutor) {
+        this.databaseContext = databaseContext;
+        this.filePreferences = filePreferences;
+        this.aiService = aiService;
+        this.taskExecutor = taskExecutor;
+
+        this.aiIngestor = new AiIngestor(aiService);
+
+        configure();
+
+        listenToPreferences();
+    }
+
+    private void configure() {
+        showToUser(true);
+        willBeRecoveredAutomatically(true);
+        updateProgress(1, 1);
+        titleProperty().set(Localization.lang("Embeddings generation"));
+
+        this.onFailure(e -> {
+            throw new RuntimeException(e);
+        });
+    }
+
+    private void listenToPreferences() {
+        // TODO: Will these listeners align with AiIngestor preference listeners?
+        aiService.getPreferences().onEmbeddingsParametersChange(this::invalidate);
+    }
+
+    public void addToProcess(Collection<BibEntry> bibEntries) {
+        bibEntries.forEach(this::addToProcess);
+    }
+
+    public void removeFromProcess(Collection<BibEntry> bibEntries) {
+        bibEntries.forEach(this::removeFromProcess);
+    }
+
+    public void addToProcess(BibEntry bibEntry) {
+        addToProcess(bibEntry.getFiles());
+    }
+
+    public void removeFromProcess(BibEntry bibEntry) {
+        removeFromProcess(bibEntry.getFiles());
+    }
+
+    public void addToProcess(List<LinkedFile> linkedFiles) {
+        linkedFiles.forEach(this::addToProcess);
+    }
+
+    public void removeFromProcess(List<LinkedFile> linkedFiles) {
+        linkedFiles.forEach(this::removeFromProcess);
+    }
+
+    public void addToProcess(LinkedFile linkedFile) {
+        if (!isBlockingNewTasks) {
+            linkedFileQueue.add(linkedFile);
+
+            synchronized (lock) {
+                if (!isRunning) {
```

Comment:
I think, a better design is to use `AtomicBoolean` instead of a self-made locking mechansim? - See also https://stackoverflow.com/a/5861966/873282.